gobblin-service/src/main/java/org/apache/gobblin/service/modules/flow/BaseFlowToJobSpecCompiler.java

```diff
@@ -149,15 +150,47 @@ public void awaitHealthy() throws InterruptedException {
     return;
   }
 
-  @Override
-  public synchronized AddSpecResponse onAddSpec(Spec addedSpec) {
-    TopologySpec spec = (TopologySpec) addedSpec;
-    log.info ("Loading topology {}", spec.toLongString());
-    for (Map.Entry entry: spec.getConfigAsProperties().entrySet()) {
-      log.info ("topo: {} --> {}", entry.getKey(), entry.getValue());
+  private synchronized  AddSpecResponse onAddTopologySpec(TopologySpec spec) {
+    log.info("Loading topology {}", spec.toLongString());
+    for (Map.Entry entry : spec.getConfigAsProperties().entrySet()) {
+      log.info("topo: {} --> {}", entry.getKey(), entry.getValue());
     }
 
-    topologySpecMap.put(addedSpec.getUri(), (TopologySpec) addedSpec);
+    topologySpecMap.put(spec.getUri(), spec);
+    return new AddSpecResponse(null);
+  }
+
+  private  AddSpecResponse onAddFlowSpec(FlowSpec flowSpec) {
+    Properties flowSpecProperties = flowSpec.getConfigAsProperties();
+    if (flowSpecProperties.containsKey(ConfigurationKeys.JOB_SCHEDULE_KEY) && org.apache.commons.lang.StringUtils.isNotBlank(
+        flowSpecProperties.getProperty(ConfigurationKeys.JOB_SCHEDULE_KEY))) {
+      try {
+        new CronExpression(flowSpecProperties.getProperty(ConfigurationKeys.JOB_SCHEDULE_KEY));
+      } catch (Exception e) {
+        log.error("invalid cron schedule: {}", flowSpecProperties.getProperty(ConfigurationKeys.JOB_SCHEDULE_KEY), e);
+        flowSpec.getCompilationErrors().add(new FlowSpec.CompilationError(0, "invalid cron schedule: " + flowSpecProperties.getProperty(ConfigurationKeys.JOB_SCHEDULE_KEY) + e.getMessage()));
+        return null;
+      }
+    }
+    String response = null;
+
+    // always try to compile the flow to verify if it is compilable
+    Dag<JobExecutionPlan> dag = this.compileFlow(flowSpec);
+    // If dag is null then a compilation error has occurred
+    if (dag != null && !dag.isEmpty()) {
+      response = dag.toString();
+    }
+    // todo: should we check quota here?
+    return new AddSpecResponse<>(response);
```

Comment:
Quota should be checked here if the flow is adhoc or set to runImmediately

Reply:
So I think the only uncertainty here would be if a user has a scheduled flow, should we store their changes permanently if their flow cannot be run immediately due to quota?
Currently, the behavior is that we reject all changes if the flow is set to run immediately and quota is overrun
If the flow is not set to run immediately, it should pass.

In other systems like ADF, the schedule of a flow and the configurations of a flow are separate. So it can be argued that we accept changes, but refuse to schedule them if their quota is overrun.

Reply:
Yup, I leave the todo hint here for the next PR after mysql quota is in place. 
I think it's easy for us to not check quota for run immediate flow here and rely on the leader to check quota and skip it. That way we can achieve the goal you described. But I'm not sure whether we want to change the behavior of the service. I would prefer to disable run immediate function and only support adhoc when we migrate adhoc flow into dag action store.

Reply:
I think runImmediately is a fairly common user use-case. We had a large scale user who would store flows with a schedule very far out, and then invoke it in an adhoc way using runImmediately. This would help them run backfills with the same configurations if needed, and also avoid concurrency issues as scheduled flows validate that they're not already running (but adhoc does not do this).

Reply:
Yeah, when we disable run immediate, we should be able to support adhoc run for scheduled flow once we move adhoc flow to be triggered by dag action store instead of spec store. But now we might want to keep the same behavior as before. 
Specifically for the use case where the schedule is far out and they want to run it immediately, I believe we want to reject the request and ask them to submit that again when quota is good